### DISCUSSION TITLE
Formatter rule ideas

### DISCUSSION THREAD

**Discussion by ysugimoto** [ysugimoto/falco#274] [Ideas]
Recently I've been working on implementing VCL formatter in falco.

Here are the underlying format rules from my thought - based on biomejs.
If anyone has a suggestion about additional formatting rules, let's discuss it.

## Basic formatting

Basically formatter formats standardized syntax - put whitespace between tokens and trim it before the semicolon.

Before:

```vcl
sub vcl_recv {
 set req.http.Foo="bar" ;
  if( req.http.Host ){
    set req.backend=F_example_com;
}   
}
```

Formatted:

```vcl
sub vcl_recv {
  set req.http.Foo = "bar";
  if (req.http.Host) {
    set req.backend = F_example_com;
  }   
}
```

---

### Indent Width

**default: 2**

Define the indent width by unit. the formatted statements/propertied will have indent by provided width with `Indent Style` character.

Before:

```vcl
sub vcl_recv {
set req.http.Foo = "bar";
}
```

Formatted (indent_width: 2):

```vcl
sub vcl_recv {
  set req.http.Foo = "bar";
}
```

---

### Indent Style

**default: space**

Define the indent style. `space` value will use whitespace `" "` character, or `tab` value will use tab `\t` character.
And this character is repeated by `Indent Width` times.

```vcl
sub vcl_recv {
set req.http.Foo = "bar";
}
```

Formatted (indent_style: space):

```vcl
sub vcl_recv {
  set req.http.Foo = "bar";
}
```

---

### Trailing Comment Width

**default: 2**

Define trailing comment space size.

```vcl
sub vcl_recv {
  set req.http.Foo = "bar";
}// trailing comment
```

Formatted (trailing_comment_width: 2):

```vcl
sub vcl_recv {
  set req.http.Foo = "bar";
}  // trailing comment
```

---

### Line Width

**default: 120**

Specify max characters for each line. The overflowed characters are printed at the next line with the same ident.

[!IMPORTANT]
Inserting line-feed is judged for each expression. It means formatter does not split in the middle of a sentence.

```vcl
sub vcl_recv {
  set req.http.Foo = "lorem" req.http.Sep "ipsum" req.http.Sep "dolor" req.http.Sep "sit" req.http.Sep "amet,";
}
```

Formatted (line_width: 80):

```vcl
sub vcl_recv {
  set req.http.Foo = "lorem" req.http.Sep "ipsum" req.http.Sep "dolor"
                     req.http.Sep "sit" req.http.Sep "amet,";
}
```

If condition example:

```vcl
sub vcl_recv {
  if (req.http.Foo == "bar" && req.http.Bar = "baz" && req.http.Baz == "foo") {
      req.http.OK = "1";
  }
}
```

Formatted (line_width: 80):

```vcl
sub vcl_recv {
  if (
      req.http.Header1 == "1" && req.http.Header2 == "2" &&
      req.http.Header3 == "3" && req.http.Header4 == "4"
  ) {
    set req.http.OK = "1";
  }
}
```

---

### Explicit Declaration Concat

**default: false**

Print string concatenation character of `+` explicitly if true.

```vcl
sub vcl_recv {
  set req.http.Foo = "lorem" "ipsum" "dolor" "sit" "amet,";
}
```

Formatted (explicit_string_concat: true):

```vcl
sub vcl_recv {
  set req.http.Foo = "lorem" + "ipsum" + "dolor" + "sit" + "amet,";
}
```

---

### Sort Declaration Property

**default: false**

If true, declaration properties will be printed with alphabetecally sorted.
This rule affects to `backend`, `director`, and `table` declarations.

```vcl
backend example {
  .connect_timeout = 1s;
  .dynamic = true;
  .port = "443";
  .host = "example.com";
  .first_byte_timeout = 30s;
  .max_connections = 500;
  .between_bytes_timeout = 30s;
  .ssl = true;
  .probe = {
    .request = "GET / HTTP/1.1" "Host: example.com" "Connection: close";
    .dummy = true;
  }
}
```

Formatted (sort_declaration_property: true):

```vcl
backend example {
  .between_bytes_timeout = 30s;
  .connect_timeout = 1s;
  .dynamic = true;
  .first_byte_timeout = 30s;
  .host = "example.com";
  .max_connections = 500;
  .port = "443";
  .ssl = true;
  .probe = {
    .dummy = true;
    .request = "GET / HTTP/1.1" "Host: example.com" "Connection: close";
  }
}
```

---

### Align Declaration Property

**default: false**

If true, declaration properties will be aligned for max field character length.
This rule affects to `backend`, `director`, and `table` declarations.

```vcl
backend example {
  .connect_timeout = 1s;
  .dynamic = true;
  .port = "443";
  .host = "example.com";
  .first_byte_timeout = 30s;
  .max_connections = 500;
  .between_bytes_timeout = 30s;
  .ssl = true;
  .probe = {
    .request = "GET / HTTP/1.1" "Host: example.com" "Connection: close";
    .dummy = true;
  }
}`,
```

Formatted (align_declaration_property: true):

```vcl
backend example {
  .connect_timeout       = 1s;
  .dynamic               = true;
  .port                  = "443";
  .host                  = "example.com";
  .first_byte_timeout    = 30s;
  .max_connections       = 500;
  .between_bytes_timeout = 30s;
  .ssl                   = true;
  .probe                 = {
    .request = "GET / HTTP/1.1" "Host: example.com" "Connection: close";
    .dummy   = true;
  }
}
```

---

### Else If

**default: false**

VCL accepts a little kind of another if statement keyword like `else if`,  `elseif`, and `elsif`, sometimes it's annoying to recognize.
If this rule set `true`, print these keyword as `else if`.

```vcl
sub vcl_recv {
  if (req.http.Foo) {
    req.http.Status = "1";
  } elseif (req.http.Bar) {
    req.http.Status = "2";
  } elsif (req.http.Baz) {
    req.http.Status = "3";
  }
}
```

Formatted (else_if: true):

```vcl
sub vcl_recv {
  if (req.http.Foo) {
    req.http.Status = "1";
  } else if (req.http.Bar) {
    req.http.Status = "2";
  } else if (req.http.Baz) {
    req.http.Status = "3";
  }
}
```

---

### Return Statement Parenthesis

**default: true**

If true, coerce surrounding parenthesis state ident of `return` statement.

```vcl
sub vcl_recv {
  return lookup;
}
```

Formatted (recommend: true):

```vcl
sub vcl_recv {
  return (lookup);
}
```

**Comment by richardmarshall** [comment]
A couple things come to mind.

## Brace placement

All the VCL I've seen has the standard of placing block statement braces on the same line as their parent statement. `if`, `sub`, `switch`, etc.  As well as placing `else` / `else if` on the same line as the closing brace of the previous block.

```vcl
sub foo
{
  if (req.http.foo)
  {
    esi;
  }
  else {
    esi;
  }
}
```

Formatted:
```vcl
sub foo {
  if (req.http.foo) {
    esi;
  } else {
    esi;
  }
}
```

## New lines within statements

Not including argument lists or conditional expressions.

```vcl
set
req.http.foo
=
"bar";

set 
req.http.foo = regsub(
  req.url,
  "foo",
  "bar"
);
```

Formatted:
```vcl
set req.http.foo = "bar";

set req.http.foo = regsub(
  req.url,
  "foo",
  "bar"
);
```

## Indent case labels

By default format the placement of case labels at the same indentation level as the outer block. This keeps code in the case blocks indented the same amount as they would be in an if-else-if chain.

**default: false**

```vcl
switch (req.http.foo) {
    case "bar":
      break;
  case "baz":
    break;
default:
  break;
}
```

Formatted (false):
```vcl
switch (req.http.foo) {
case "bar":
  break;
case "baz":
  break;
default:
  break;
}
```

Formatted (true):
```vcl
switch (req.http.foo) {
  case "bar":
    break;
  case "baz":
    break;
  default:
    break;
}
```

Will look through various VCL files and think on this some more.

**Reply by ysugimoto** [reply]
These are supported(coerced) by AST-based formatting 👍

**Comment by ysugimoto** [comment]
There are two additional formatting rules we have.

## Sort Declarations

**default:false**

If true, sort root declarations by VCL directive order:

1. import/include statements
2. ACL
3. Backend
4. Director
5. Table
6. vcl_recv
7. vcl_hash
8. vcl_hit
9. vcl_miss
10. vcl_pass
11. vcl_fetch
12. vcl_error
13. vcl_deliver
14. vcl_log
15. user-defined subroutines

```vcl
sub vcl_fetch {
  ...
}

sub vcl_deliver {
  ...
}

sub vcl_recv {
  ...
}
```

Formatted (recommend: true):

```vcl
sub vcl_recv {
  ...
}

sub vcl_fetch {
  ...
}

sub vcl_deliver {
  ...
}
```

---

## Align Trailing Comment

**default:false**

If true, align the trailing comment for each statement/declaration line:

```vcl
sub vcl_recv {
  set req.http.X-Forwarded-Host = "example.com";  // request comes from example.com
  set req.backend = example_com;  // set backend
}
```

Formatted (align_trailing_comment: true):


```vcl
sub vcl_recv {
  set req.http.X-Forwarded-Host = "example.com";  // request comes from example.com
  set req.backend = example_com;                  // set backend
}
```

---

## Always Next Line ElseIf

**default:false**

If true, the `else if` and `else` statement prints next line.

```vcl
sub vcl_recv {
  if (req.http.Foo) {
    ...
  } else if (req.http.Bar) {
    ...
  } else {
    ...
  }
}
```

Formatted (true):

```vcl
sub vcl_recv {
  if (req.http.Foo) {
    ...
  }
  else if (req.http.Bar) {
    ...
  }
  else {
    ...
  }
}
```
